src/denoise.rs

```diff
@@ -0,0 +1,607 @@
+      *sw = Self::spatial_window(j as f64 + 0.5);
+    });
+    Self::normalize_for_overlap_add(&mut sw);
+
+    let nscale = 1.0 / (BLOCK_VOLUME as f64).sqrt();
+    for j in 0..TB_SIZE {
+      for k in 0..SB_SIZE {
+        for q in 0..SB_SIZE {
+          hw[(j * SB_SIZE + k) * SB_SIZE + q] =
+            (tw[j] * sw[k] * sw[q] * nscale) as f32;
+        }
+      }
+    }
+
+    hw
+  }
+
+  #[inline(always)]
+  // Hanning windowing
+  fn spatial_window(n: f64) -> f64 {
+    0.5 - 0.5 * (2.0 * PI * n / SB_SIZE as f64).cos()
+  }
+
+  #[inline(always)]
+  // Simple rectangular windowing
+  fn temporal_window() -> f64 {
+    1.0
+  }
+
+  // Accounts for spatial block overlap
+  fn normalize_for_overlap_add(hw: &mut [f64]) {
+    let inc = SB_SIZE - SO_SIZE;
+
+    let mut nw = [0f64; SB_SIZE];
+    let hw = &mut hw[..SB_SIZE];
+
+    for q in 0..SB_SIZE {
+      for h in (0..=q).rev().step_by(inc) {
+        nw[q] += hw[h].powi(2);
+      }
+      for h in ((q + inc)..SB_SIZE).step_by(inc) {
+        nw[q] += hw[h].powi(2);
+      }
+    }
+
+    for q in 0..SB_SIZE {
+      hw[q] /= nw[q].sqrt();
+    }
+  }
+
+  #[inline]
+  unsafe fn proc0(
+    &self, s0: &[T], s1: &[f32], dest: &mut [f32], p0: usize, p1: usize,
+    src_scale: f32,
+  ) {
+    let s0 = s0.as_ptr();
+    let s1 = s1.as_ptr();
+    let dest = dest.as_mut_ptr();
+
+    for u in 0..p1 {
+      for v in 0..p1 {
+        let s0 = s0.add(u * p0 + v);
+        let s1 = s1.add(u * p1 + v);
+        let dest = dest.add(u * p1 + v);
+        dest.write(u16::cast_from(s0.read()) as f32 * src_scale * s1.read())
+      }
+    }
+  }
+
+  #[inline]
+  unsafe fn proc1(
+    &self, s0: &[f32], s1: &[f32], dest: &mut [f32], p0: usize, p1: usize,
+  ) {
+    let s0 = s0.as_ptr();
+    let s1 = s1.as_ptr();
+    let dest = dest.as_mut_ptr();
+
+    for u in 0..p0 {
+      for v in 0..p0 {
+        let s0 = s0.add(u * p0 + v);
+        let s1 = s1.add(u * p0 + v);
+        let dest = dest.add(u * p1 + v);
+        dest.write(s0.read().mul_add(s1.read(), dest.read()));
+      }
+    }
+  }
+
+  #[inline]
+  fn remove_mean(
+    &self, dftc: &mut [Complex<f32>; COMPLEX_COUNT],
+    dftgc: &[Complex<f32>; COMPLEX_COUNT],
+    means: &mut [Complex<f32>; COMPLEX_COUNT],
+  ) {
+    let gf = dftc[0].re / dftgc[0].re;
+
+    for h in 0..COMPLEX_COUNT {
+      means[h].re = gf * dftgc[h].re;
+      means[h].im = gf * dftgc[h].im;
+      dftc[h].re -= means[h].re;
+      dftc[h].im -= means[h].im;
+    }
+  }
+
+  #[inline]
+  fn add_mean(
+    &self, dftc: &mut [Complex<f32>; COMPLEX_COUNT],
+    means: &[Complex<f32>; COMPLEX_COUNT],
+  ) {
+    for h in 0..COMPLEX_COUNT {
+      dftc[h].re += means[h].re;
+      dftc[h].im += means[h].im;
+    }
+  }
+
+  #[inline]
+  // Applies a generalized wiener filter
+  fn filter_coeffs(&self, dftc: &mut [Complex<f32>; COMPLEX_COUNT]) {
+    for h in 0..COMPLEX_COUNT {
+      let psd = dftc[h].re.mul_add(dftc[h].re, dftc[h].im.powi(2));
```

Comment:
Is this missing squaring the real part?

Reply:
No, it translates to `real * real + im^2`. The reason it's done like this is because `mul_add` is more efficient and accurate for floating point math.